docs/storage.md

```diff
@@ -0,0 +1,187 @@
+# Ingestor WAL Storage Management and Backpressure
+
+## Overview
+
+The Ingestor component of ADX-Mon is responsible for reliably buffering, batching, and uploading telemetry data (metrics, logs, traces) to Azure Data Explorer (ADX). To ensure durability and high throughput, Ingestor uses a Write-Ahead Log (WAL) mechanism, writing all incoming data to disk before upload. This document details how Ingestor manages WAL files, prevents storage exhaustion, and applies backpressure to maintain system stability.
+
+---
+
+## WAL Segment Lifecycle
+
+### 1. WAL Segments per Table/Schema
+- Each unique table/schema combination gets its own WAL segment file.
+- Incoming telemetry is appended to the appropriate segment.
+
+### 2. Segment Rotation
+- Segments are rotated (closed and a new one started) when either:
+  - **Max Segment Size** is reached (e.g., 100MB–1GB, configurable)
+  - **Max Segment Age** is reached (e.g., 5 minutes, configurable)
+- Rotation is triggered both on write and by a background goroutine (periodic check).
+
+### 3. Segment Handling After Rotation
+- **Small segments** (below a threshold, e.g., <100MB) are transferred to a peer Ingestor for coalescing.
+- **Large or old segments** are uploaded directly to ADX.
+- If peer transfer fails, the segment is uploaded directly.
+
+### 4. Upload and Cleanup
+- Segments are compressed and uploaded in batches to ADX.
+- After successful upload or transfer, segments are deleted from local storage.
+
+---
+
+## Storage Limits and Backpressure
+
+### Configurable Limits
+- **Max Disk Usage**: Total bytes allowed for all WAL segments, including:
+  - All closed segments (on disk)
+  - The active (open) segment
+  - Any in-flight (not yet flushed) writes buffered in memory
+- **Max Segment Count**: Maximum number of WAL segments allowed on disk.
+
+### Enforcement
+- On every write, Ingestor checks if adding data would exceed disk usage or segment count limits.
+- **Disk usage enforcement is comprehensive:**
```

Comment:
Sort of true. Many of the metrics like current index size are populated async on a timer. There can be gaps currently between when a write has finished (so it's removed from currentInflight) and when the async process updates it.